### PR TITLE
TTL improvements

### LINKED ISSUE #47: Different number of rising/falling edges
Hi there, I'm trying to add some TTL traces from CED into and existing nwb file with a notebook from I think @CodyCBakerPhD . I run into the following issue.

image
image

When ignoring this line, the code runs, but I'm wondering how vital this line is, and whether the resulting file will make sense without it.

Any help would be much appreciated!

All the best

Ross

### PR DESCRIPTION
**Pull request by lkeegan**
key changes:

- https://github.com/catalystneuro/mease-lab-to-nwb/commit/f15e03dcd2be591816c73d4da43f2f6045348fc0 add heuristic to check if TTL data is just noise
  - resolves #47 
- 60f604da546187c8d2e152e0c9a8a081dbefd3da write TTL start/stop times as IntervalSeries in stimulus instead of as TimeIntervals in processing
  - also fixes issue where times were truncated to ints (https://github.com/catalystneuro/mease-lab-to-nwb/blob/5da0994e90996f0e597f4e150aff192a6994a5e1/mease_lab_to_nwb/convert_ced/cedstimulusinterface.py#L59)
- https://github.com/catalystneuro/mease-lab-to-nwb/commit/0dc4ba8e717336e78b27fd6ff6e453b2527b3a86 add Laser channel as OptogeneticSeries: (dummy Device and OptogeneticStimulusSite for now)

### CODE REVIEW

File: tests/test_CEDNWBConverter.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from mease_lab_to_nwb import CEDNWBConverter
+import pynwb
+from pathlib import Path
+import pytest
+import os
+
+
+def test_cednwbconverter(tmp_path):
+    # read smrx file
+    file_recording = str((Path(__file__).parent / "data" / "TTLtest.smrx").resolve())
```
Comment (CodyCBakerPhD) [inline tests/test_CEDNWBConverter.py:10]: It looks like you're used to using pretty sophisticated workflows with cached info - if interested, you can actually setup the CI to handle some test CED data as well; see [this workflow](https://github.com/catalystneuro/nwb-conversion-tools/blob/master/.github/workflows/update-testing-data.yml) we use for downloading/caching a test file from the test dataset to use within the [CI tests](https://github.com/catalystneuro/nwb-conversion-tools/blob/master/tests/test_gin.py) that reference the data within that environment.

**Comment by CodyCBakerPhD** [comment]
@lkeegan Thanks for the lovely code!

While it's fine to explicitly specify the `IntervalSeries` timestamps field, note that you may also be able to use the `add_interval` function (https://github.com/catalystneuro/nwb-conversion-tools/blob/master/tests/test_gin.py) in a bit more direct fashion.

Also, if you could remind me what issue you were having with `TimeIntervals`? The times there should be supported as floats equivalently.

Otherwise, if the NWBFiles output by these changes look good to you, then can merge to stay up to date!

**Review by CodyCBakerPhD — APPROVED** [review]

File: tests/test_CEDNWBConverter.py (hunk 1 above)
Comment (lkeegan) [reply tests/test_CEDNWBConverter.py:10]: @CodyCBakerPhD thanks! I'll try this out. Should I also remove the smrx test file I added here and open a PR to add it to https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/src/master/spike2 instead?

**Comment by lkeegan** [comment]
> While it's fine to explicitly specify the `IntervalSeries` timestamps field, note that you may also be able to use the `add_interval` function (https://github.com/catalystneuro/nwb-conversion-tools/blob/master/tests/test_gin.py) in a bit more direct fashion.
> 
Thanks! Have changed to using this instead: https://github.com/catalystneuro/mease-lab-to-nwb/pull/48/commits/819e8c6a407693f02563f761645e8c8507f20ad9

> Also, if you could remind me what issue you were having with `TimeIntervals`? The times there should be supported as floats equivalently.
> 
I think the issue was that the timestamps were being generated from tick counts divided by sampling frequency, and as both of these were integers it was doing integer division instead of floating point i.e. the times all got rounded down to the nearest second (before being passed to the `TimeIntervals` object):

https://github.com/catalystneuro/mease-lab-to-nwb/blob/5da0994e90996f0e597f4e150aff192a6994a5e1/mease_lab_to_nwb/convert_ced/cedstimulusinterface.py#L61

File: tests/test_CEDNWBConverter.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply tests/test_CEDNWBConverter.py:11]: They would _love_ some `.smrx` files! That's currently a gap in both our coverage @samuelgarcia and @JuliaSprenger (the managers of that repo) would be happy to help get some in - the only thing I'd mention is it's critical for CI workflows that the example file that gets uploaded is very small (< 10 MB, ideally) while still containing all relevant header info + a snippet of data. That's *usually* comparable to a minute or less of recording, or simply an existing recording with all the data after that point clipped off.

**Comment by CodyCBakerPhD** [comment]
Going to go ahead and merge this so the PR doesn't get too clogged - once the GIN stuff is worked out we can add that in later.

Thanks for the code, and for opening the PR here so we can stay up to date!

File: tests/test_CEDNWBConverter.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply tests/test_CEDNWBConverter.py:11]: Ahh, I see it - wasn't looking under `spike2`! Thanks Sam

**Pull request merged**